text/0000-light-dom.md

```diff
@@ -0,0 +1,218 @@
+
+To support the cases where a shadow DOM element composes a light element, light DOM styles are required to be injected to the closest root node. For a given light DOM elements if all the ancestor components are also a light DOM components, the component style sheet will be injected in the document `<head>`. Otherwise if any of the ancestors is a shadow DOM components the style has to be injected in the closest shadow root. Upon insertion of a light DOM element does the following steps:
+
+- look up for the closest root node (`Node.prototype.getRootNode()`)
+- insert the stylesheet if not already present:
+  - if the root node is the HTML document, the style sheet is inserted as a `<style>` element in the `<head>`.
+  - if the root node is a shadow root, the stylesheet is inserted as a `<style>` element as the first shadow root child.
+
+It is important to notice that the order in which light DOM components are rendered impact the order in which stylesheets are injected in the root node and directly influences CSS rule specificity.
+
+**Note:** Different approaches layer style scoping on top has been discussed while designing Light DOM, like introducing a new file extension for automatic style scoping `.scoped.css` or using a `<style scoped>` element in the template.
+
+#### `this.template`
+
+In `LightningElement`, `this.template` returns the shadow-root. It will return `null` in `MacroElement`.
+
+### Querying
+
+`MacroElement` (and `LightningElement`) forward several DOM querying/manipulation methods like `querySelector`, `dispatchEvent` etc. to the host element. Full list of methods forwarded [here](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine-core/src/framework/base-lightning-element.ts#L242).
+
+This will allow component author to just use `this.querySelector` instead of using `this.template.querySelector` in case of `LightningElement`.
+However, there's no way to get a direct reference to the host element, like `this.template.host` that's available in `LightningElement`.
+
+E.g.
+
+```js
+/* <template>
+ *   <x-child></x-child>
+ * </template>
+ */
+export default class ParentElement extends MacroElement {
+  renderedCallback() {
+    this.querySelector("x-child").addEventListener("custom", (e) => {
+      console.log("Called with " + JSON.stringify(e.detail));
+    });
+  }
+}
+
+/**
+ * <template>
+ *   <button>Click Me</button>
+ * </template>
+ */
+export default class ChildElement extends MacroElement {
+  renderedCallback() {
+    this.querySelector("button").addEventListener("click", this.handleClick);
+  }
+
+  handleClick() {
+    this.dispatchEvent(new CustomEvent("custom", { detail: { macro: true } }));
+  }
+}
+```
+
+### Security (WIP)
+
+- In some applications, light-dom components may not be allowed... it’s up to the app context
```

Comment:
Do we have an example of how light DOM components can be disabled? Will there be a compiler flag to accomplish this?